compiler/ccgexprs.nim

```diff
@@ -981,16 +981,17 @@ proc genEcho(p: BProc, n: PNode) =
   # is threadsafe.
   internalAssert p.config, n.kind == nkBracket
   if p.config.target.targetOS == osGenode:
-    # bypass libc and print directly to the Genode LOG session
+    # echo directly to the Genode LOG session
     var args: Rope = nil
     var a: TLoc
     for it in n.sons:
       if it.skipConv.kind == nkNilLit:
         add(args, ", \"\"")
       else:
         initLocExpr(p, it, a)
-        add(args, ropecg(p.module, ", #nimToCStringConv($1)", [rdLoc(a)]))
+        add(args, ropecg(p.module, ", Genode::Cstring($1->data, $1->len)", [rdLoc(a)]))
```

Comment:
All this `p.config.target.targetOS == osGenode` feels awkward in the compiler code. Why isn't genode support inside `echoBinSafe` impl instead?

Reply:
It is awkward, but as I recall this is the worst of it. Its important that the base implementation of echo should use the logging RPC because stdio descriptors are not native on Genode, those are implemented by the POSIX compatibility layer. I've tried to move this code out of this file and into the stdlib but I had issues possibly related to `{.importcpp.}`.